File: CaloriesTrackerIOS/Domain/Units/UnitConverter.swift
import Foundation

enum UnitConverterError: Error, LocalizedError {
    case invalidAmount
    case incompatibleServingUnit

    var errorDescription: String? {
        switch self {
        case .invalidAmount:
            "Количество должно быть конечным неотрицательным числом."
        case .incompatibleServingUnit:
            "Дополнительная единица несовместима с базовой единицей продукта."
        }
    }
}

enum UnitConverter {
    /// Converts an amount in a historical serving unit to the base-unit amount of its ProductVersion.
    static func baseAmount(
        from amount: Double,
        servingUnit: ServingUnit,
        expectedBaseUnit: ProductBaseUnit,
    ) throws -> Double {
        guard amount.isFinite, amount >= 0,
              servingUnit.conversionAmount.isFinite, servingUnit.conversionAmount > 0
        else {
            throw UnitConverterError.invalidAmount
        }

        guard servingUnit.conversionUnit.rawValue == expectedBaseUnit.rawValue else {
            throw UnitConverterError.incompatibleServingUnit
        }

        return amount * servingUnit.conversionAmount
    }
}

File: web/src/domain/units/unit-converter.ts
import type { ProductVersion } from '@/domain/products/product-version'
import type { ServingUnit } from '@/domain/products/serving-unit'

/** Converts a selected serving unit without calculating nutrition. */
export class UnitConverter {
  convertServingAmount(amount: number, servingUnit: ServingUnit): ConvertedAmount {
    if (!Number.isFinite(amount) || amount < 0) {
      throw new Error('Amount must be a non-negative finite number.')
    }

    if (!Number.isFinite(servingUnit.conversionAmount) || servingUnit.conversionAmount <= 0) {
      throw new Error('Serving conversion amount must be a positive finite number.')
    }

    return {
      amount: amount * servingUnit.conversionAmount,
      unit: servingUnit.conversionUnit,
    }
  }

  toBaseAmount(productVersion: ProductVersion, amount: number, servingUnit: ServingUnit): number {
    const convertedAmount = this.convertServingAmount(amount, servingUnit)

    if (productVersion.baseUnitType === 'serving' || convertedAmount.unit !== productVersion.baseUnitType) {
      throw new Error('Serving unit is incompatible with the product base unit.')
    }

    return convertedAmount.amount
  }
}

export const unitConverter = new UnitConverter()

export interface ConvertedAmount {
  amount: number
  unit: ServingUnit['conversionUnit']
}
